### PR TITLE
Add organisation

### PR DESCRIPTION
**Pull request by Vampouille**

### CODE REVIEW

**Comment by Vampouille** [comment]
Work in progress, please do not merge !

**Comment by Vampouille** [comment]
Work is finished, please review

File: ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp [hunk 1]
```diff
@@ -47,7 +47,7 @@ try {
     <meta charset="UTF-8">
     <title>LDAPadmin - <%= instanceName %></title>
     <link rel="stylesheet" href="lib/bootstrap/css/bootstrap.min.css" />
-
+    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/select2/4.0.3/css/select2.css"/>
```
Comment (fvanderbiest) [inline ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp]: We voted against CDNs on https://github.com/georchestra/georchestra/issues/939 
I'd prefer to have this resource locally instead (same thing below)

I would not be opposed to your PR fixing this other issue at the same time.

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ds/AccountDaoImpl.java [hunk 2]
```diff
@@ -563,6 +587,32 @@ public Object mapFromContext(Object ctx) {
...
+                    String group = it.next();
+                    Matcher m = this.pattern.matcher(group);
+
+                    // Skip roles
+                    if (!m.matches())
+                        continue;
+
+                    // Check organization cardinality
+                    if (org != null)
+                        throw new RuntimeException("More than one org per user on " + account.getCommonName());
```
Comment (fvanderbiest) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ds/AccountDaoImpl.java]: :+1:

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ds/GroupDaoImpl.java [hunk 3]
```diff
@@ -70,8 +71,8 @@
 
 	private String uniqueNumberField = "ou";
 
-    private LdapRdn groupSearchBaseDN;
-    private LdapRdn userSearchBaseDN;
+    private Name groupSearchBaseDN;
+	private LdapRdn userSearchBaseDN;
```
Comment (fvanderbiest) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ds/GroupDaoImpl.java]: indent ?

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/utils/ResponseUtil.java [hunk 4]
```diff
@@ -52,14 +52,17 @@ public static String buildResponseMessage(Boolean status){
...
+		try {
+			res.put("success", status);
+			if(errorMessage != null){
+				res.put("error", errorMessage);
+			}
+		} catch (JSONException e) {
+			throw new RuntimeException(e);
 		}
-		return error;
+		return res.toString();
```
Comment (fvanderbiest) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/utils/ResponseUtil.java]: thanks for the rewrite :+1:

File: ldapadmin/src/main/webapp/WEB-INF/views/createAccountForm.jsp [hunk 5]
```diff
@@ -192,6 +195,7 @@
         $("input#uid").attr("placeholder", "<s:message code="uid.placeholder" />");
         $("input#password").attr("placeholder", "<s:message code="password.placeholder" />");
         $("input#confirmPassword").attr("placeholder", "<s:message code="confirmPassword.placeholder" />");
+		$("#org").select2();
```
Comment (fvanderbiest) [inline ldapadmin/src/main/webapp/WEB-INF/views/createAccountForm.jsp]: indent

**Comment by fvanderbiest** [comment]
Nice job !

Shouldn't we also PR on https://github.com/georchestra/datadir/blob/master/ldapadmin/ldapadmin.properties and https://github.com/georchestra/georchestra/blob/15.12/config/defaults/ldapadmin/maven.filter ?

File: ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp (hunk 1 above)
Comment (Vampouille) [inline ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp]: fixed by 93991c6

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/utils/ResponseUtil.java [hunk 6]
```diff
@@ -52,14 +52,17 @@ public static String buildResponseMessage(Boolean status){
 	}
 	
 	public static String buildResponseMessage(Boolean status, String errorMessage){
-		
-		String error;
-		if(errorMessage == null){
-			error = "{ \"success\": "+ status.toString() +" }";
-		} else {
-			error = "{ \"success\": "+ status.toString() +" , \"error\": \"" +errorMessage+ "\" }";
```
Comment (pmauduit) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/backoffice/utils/ResponseUtil.java]: :+1:

File: ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/newaccount/NewAccountFormController.java [hunk 7]
```diff
@@ -137,9 +148,18 @@ public String setupForm(HttpServletRequest request, Model model) throws IOExcept
...
-						 SessionStatus sessionStatus)
+						 SessionStatus sessionStatus,
+						 Model model)
 			throws IOException {
 
+		// Populate orgs droplist
+		Map<String, String> options = new HashMap<String, String>();
+		List<Org> orgs = this.orgDao.findAll();
+		for(Org org : orgs)
+			options.put(org.getId(), org.getName());
```
Comment (pmauduit) [inline ldapadmin/src/main/java/org/georchestra/ldapadmin/ws/newaccount/NewAccountFormController.java]: indent ?

**Comment by pmauduit** [comment]
lgtm

**Comment by fvanderbiest** [comment]
Also missing `shared.ldap.orgsSearchBaseDN = ou=orgs` in  https://github.com/georchestra/template/blob/15.12/build_support/shared.maven.filters#L87-L91 and https://github.com/georchestra/georchestra/blob/15.12/config/shared.maven.filters#L14-L24

File: ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp [hunk 8]
```diff
@@ -140,7 +140,8 @@ try {
     <script src="js/filters.js"></script>
     <script src="js/directives.js"></script>
     <script src="lib/bootstrap/js/bootstrap.min.js"></script>
-    <script type="text/javascript" src="//cdn.jsdelivr.net/underscorejs/1.4.3/underscore-min.js"></script>
-    <script type="text/javascript" src="//cdn.jsdelivr.net/restangular/1.5.1/restangular.min.js"></script>
+    <script type="text/javascript" src="lib/underscorejs/underscore-min.js"></script>
+    <script type="text/javascript" src="lib/restangular/restangular.min.js"></script>
+    <script type="text/javascript" src="lib/select2/select2.full.js"></script>
   </body>
```
Comment (fvanderbiest) [inline ldapadmin/src/main/filtered-resources/WEB-INF/views/privateUi.jsp]: Two other resources in this file are loaded from a CDN:
```
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.10.1/jquery.min.js"></script>
    <script src="//ajax.googleapis.com/ajax/libs/angularjs/1.1.5/angular.min.js"></script>
```

**Comment by Vampouille** [comment]
PR for template : https://github.com/georchestra/template/pull/29

**Comment by fvanderbiest** [comment]
Once you're done, feel free to merge the 3 PRs. 
Thanks again !

**Comment by fvanderbiest** [comment]
(and also to cascade the changes from master to docker branches in the datadir and template repos)

**Comment by Vampouille** [comment]
https://github.com/Vampouille/georchestra/blob/Add_Organisation/ldapadmin/src/main/webapp/WEB-INF/views/createAccountForm.jsp#L125
Should I also retrieve this resource ?

**Pull request merged**